### PR TITLE
Document w3i versions 0-31

### PR DESCRIPTION
**Pull request by Luashine**
From the commit message:

----

While this should be just an update to the previous document,
the diff is very large due to Markdown tables and I used a different
converter this time.

Changes:
- All known versions described, 0-31
- Added Version History section
- For string paths, specified the max allowed length
- Some minor corrections and rewordings

:tada: :partying_face: :tada: Sorry for making all your parsers and tools obsolete :)

### CODE REVIEW

**Comment by WaterKnight** [comment]
Does the "Max X chars" mean that after up to X chars there cannot be any more bytes for this node or can there be more bytes but the string value this represents will be truncated? Does X include the \0 terminator? Or can the last character be another character than the \0 terminator?

**Comment by WaterKnight** [comment]
Unit and item ids can be special in unit and item tables:

> iDNR = random item, uDNR = random unit

> a unit/item id of 0x00000000 indicates that no unit/item is created

https://867380699.github.io/blog/2019/05/09/W3X_Files_Format#war3mapunitsdoo

**Comment by Luashine** [comment]
> Unit and item ids can be special in unit and item tables:

I am not sure it belongs here, Hodor had it documented as part of the .doo file. If this is something more common then it should be documented outside of a single format document. Either provide an example or this is something to be checked during a later data mining + spec verification process.

> Does the "Max X chars"

I need to follow up. It should be the C buffer size incl. the `\0` terminator, that's how the `MAX_PATH` [is used](https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation).

There are two types of lengths for strings:

- (THIS) The string reader has an explicit size limit specified
- (rare) The string type is preceded by a string length integer. WorldEdit seems to be reading garbage from memory in that case, basically `memcpy` without parsing from str[0] to the first occurrence of `\0`

The wording will need to change for this one. Something like `string_max(260)`?

**Comment by WaterKnight** [comment]
The chunks are players are optional, at least in the common versions, see the eof checks in https://github.com/inwc3/wc3libs/blob/c86fede635cd46bca074a3280111baf44e54ce2e/src/main/java/net/moonlightflower/wc3libs/bin/app/W3I.java#L2620

**Comment by WaterKnight** [comment]
> I am not sure it belongs here, Hodor had it documented as part of the .doo file. If this is something more common then it should be documented outside of a single format document. Either provide an example or this is something to be checked during a later data mining + spec verification process.

Yes, since it's used in multiple places, it can be referenced instead. But it's not the case everywhere, so it should be marked imo.

> The wording will need to change for this one. Something like string_max(260)?

Maybe, but I think it can just be mentioned globally somewhere and then stick to a uniform format where it is like this.

**Comment by Luashine** [comment]
> The chunks after players are optional, at least in the common versions, see the eof checks in

You are right on this one, but the game does not determine it based on EOF instead it's a switch with multiple defined levels. Mimicking HTTP, it may be called Trailer Data Level:

0. Read nothing after Game Data Version
1. Read player Data
2. Read Team Data (forces)
3. Read Custom Upgrades, Tech, Unit Tables, Item Tables, Format's 26,27 Scripting language (i.e. this is the final level as of Reforged)

The format version rules still apply. The Trailer Data Level is apparently determined before any parsing of W3I... I wonder if it has anything to do with saves or campaigns?

Because there's a dirty little secret: the game's parsers are literally raw memory serializers/deserializers and are applying side-effects in-flight, without a clear separation between parsing & interpretation steps.

I'd consider EOF detection is a workaround to get it working but I'd write it down as `IF( TRAILER_LEVEL >= n )` for the time being until the logic is understood clearly. Do you agree? And with the wording?

**Comment by Luashine** [comment]
In-depth testing about max chars: https://github.com/Luashine/wc3-test-maps/tree/master/SuggestedPlayerInW3I

TLDR: Max buffer in bytes, must stay under this size (incl. `\0`). Spec will be updated and pushed later, I got the unknown types wrong for the 6 int/floats too.

**Comment by ChiefOfGxBxL** [comment]
This looks good. I especially like the detailed version table at the end.

For consistency, you may want to refer to the format version as just `Format <= 8`
